### ISSUE TITLE
ASP.NET Core integration only reads settings from env vars

### ISSUE THREAD

**Issue by andreas-soroko** [DataDog/dd-trace-dotnet#632]
**Describe the bug**
I hoped that the v1.12 will fix that issue but it doesnt 
 
i dont know, maybe we are doing something wrong but i dont think so because the integration for http (out)/database calls still works 

**Runtime environment (please complete the following information):**
- Instrumentation mode: automatic 
- Tracer version: 1.12
- OS: mcr.microsoft.com/dotnet/core/runtime-deps:3.1
- CLR: .NET Core 3.1

**Additional context**
 **Partial parts from Docker image:** 
```
FROM                    mcr.microsoft.com/dotnet/core/runtime-deps:3.1
....
#Install Datadog APM packages
ARG DD_VERSION=1.12.0
RUN mkdir -p /tmp
RUN curl -L https://github.com/DataDog/dd-trace-dotnet/releases/download/v${DD_VERSION}/datadog-dotnet-apm_${DD_VERSION}_amd64.deb --output /tmp/datadog_apm.deb
RUN sudo dpkg -i /tmp/datadog_apm.deb


#Setup Datadog APM
ENV CORECLR_ENABLE_PROFILING=1
ENV CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
ENV CORECLR_PROFILER_PATH=/opt/datadog/Datadog.Trace.ClrProfiler.Native.so
ENV DD_INTEGRATIONS=/opt/datadog/integrations.json
ENV DD_DOTNET_TRACER_HOME=/opt/datadog
```
> note: installs without any problems 

**Trace initialization**

```
Tracer.Instance = Tracer.Create(
    new Uri($"http://{AWSEnvironment.DockerHost}:8126"),
    <serviceName>,
    true
);
``` 
> note: isDebugEnabled: true, there is no debug file in the directory ( in the container ) 

**Partial parts from Startup.cs**
```
        public virtual void ConfigureServices(IServiceCollection services)
        {
            // ....
            services.AddControllers()
                .AddNewtonsoftJson(options =>
                {
                    options.SerializerSettings.ContractResolver = new Newtonsoft.Json.Serialization.CamelCasePropertyNamesContractResolver();
                    options.SerializerSettings.NullValueHandling = NullValueHandling.Ignore;
                    options.SerializerSettings.Formatting = Formatting.None;
                })
                .AddListTypeConverter()
                .AddMvcOptions(options =>
                {
                    options.InputFormatters.Add(new TextPlainInputFormatter());
                    options.OutputFormatters.Add(new HtmlOutputFormatter());
                    options.Conventions.Add(new AcceptHeaderToProducesMapper());
                    options.Conventions.Add(new LowerControllerModelConvention());
                });
                // ....
        }

        public virtual void Configure(IApplicationBuilder app)
        {
            // ....
            app.UseRouting();
            app.UseEndpoints(endpoints => { endpoints.MapControllers(); });
            // ....
        }
```

**Comment by stevejgordon** [comment]
This seems to be an issue for us too. Our service is a very lightweight ASP.NET Core application which uses two custom middleware components.

```csharp
app.UseMiddleware<DataDogMetricsMiddleware>();

app.MapWhen(ctx => ctx.Request.Method == GetMethod
	&& (ctx.Request.Path.StartsWithSegments(new PathString("/somepath")), appBuilder =>
	{
		appBuilder.UseMiddleware<CustomImageResponseMiddleware>();
	});
```

We have added APM to our Dockerfile as follows...

```
FROM mcr.microsoft.com/dotnet/core/aspnet:3.1 AS base
WORKDIR /app
EXPOSE 80
ENV DOTNET_SKIP_FIRST_TIME_EXPERIENCE=true DOTNET_CLI_TELEMETRY_OPTOUT=1

# Install DataDog APM
RUN curl -LO https://github.com/DataDog/dd-trace-dotnet/releases/download/v1.12.0/datadog-dotnet-apm_1.12.0_amd64.deb
RUN dpkg -i ./datadog-dotnet-apm_1.12.0_amd64.deb

# Setup DataDog APM
ENV CORECLR_ENABLE_PROFILING=1 
ENV DD_LOGS_INJECTION=true 
ENV CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8} 
ENV CORECLR_PROFILER_PATH=/opt/datadog/Datadog.Trace.ClrProfiler.Native.so 
ENV DD_INTEGRATIONS=/opt/datadog/integrations.json 
ENV DD_DOTNET_TRACER_HOME=/opt/datadog 
ENV DD_DIAGNOSTIC_SOURCE_ENABLED=true 

...
```

We configure the APM from the Program.cs by calling...

```csharp
private static void ConfigureDataDogApm(IConfiguration configuration)
{
	if (!configuration.GetValue("DataDog:ApmEnabled", false))
		return;

	var dataDogAgentServer = EcsHostInfo.HostIpV4;

	var settings = TracerSettings.FromDefaultSources();

	settings.ServiceName = configuration.GetValue<string>("DataDog:ApmServiceName");
	settings.AgentUri = new Uri($"http://{dataDogAgentServer}:{configuration.GetValue<string>("DataDog:ApmPort")}", UriKind.Absolute);

	Tracer.Instance = new Tracer(settings);
}
```

Our container is hosted on AWS ECS.

When I view the results, I can see the collection of the http-client and redis integrations for outbound requests. However, I do not see a general request span for each request as a whole. I was expecting this to be collected via the DiagnosticSource integration.

I can achieve this by manually creating a span in my middleware...

```csharp
using (var tracerScope = Tracer.Instance.StartActive("web.request"))
{
	var host = context.Request.Host.ToString();
	var httpMethod = context.Request.Method.ToUpperInvariant();
	var url = context.Request.GetDisplayUrl().ToLowerInvariant();
	var path = UriHelper.BuildRelative(context.Request.PathBase, context.Request.Path, context.Request.QueryString);

	var span = tracerScope.Span;

	span.Type = SpanTypes.Web;
	span.ResourceName = $"{httpMethod} {path.ToLowerInvariant()}";

	span.SetTag(Tags.HttpMethod, httpMethod);
	span.SetTag(Tags.HttpUrl, path);
	span.SetTag(Tags.HttpRequestHeadersHost, host);
	span.SetTag(Tags.InstrumentedMethod, nameof(DataDogMetricsMiddleware));
	span.SetTag(Tags.Language, "dotnet");
	span.SetTag(Tags.SpanKind, SpanKinds.Server);

	// DO STUFF
}
```

**Comment by AndreyMarchuk** [comment]
Sounds like you're having the same issue as https://github.com/DataDog/dd-trace-dotnet/issues/581
Fix is going to be a part of v1.12.1

**Comment by ffandreassoroko** [comment]
> Fix is going to be a part of v1.12.1

When will v1.12.1 be published ?

**Comment by lucaspimentel** [comment]
@AndreyMarchuk: that issue you linked to is about ADO.NET traces, not ASP.NET Core.

**Comment by stevejgordon** [comment]
This is still an issue for me when using 1.13.0.

**Comment by marcin-dardzinski** [comment]
I have a similar problem when I'm using automatic instrumentation. I can see a trace for a http request, I can also see traces for sql and http client calls made during handling of this request, but they are uncorrelated - I can't see sql traces as children of a web request.

I ended up we a solution similar to stevejgordon - disabling automatic AspNetCore instrumentation and creating a custom middleware to open web request scope. Then I could see traces correlated correctly.

I am using ASP.NET Core 3.1 on mcr.microsoft.com/dotnet/core/aspnet:3.1 Docker image with 1.13.0 tracer version

**Comment by stevejgordon** [comment]
Is there any update on this?

**Comment by chris-sargood** [comment]
Same issue. I'm using microsoft/dotnet:2.2 and after upgrading to v1.13.2 the aspnet_coremvc.request span seems to have gone missing

**Comment by jonjomckay** [comment]
I'm not sure if it's related, but we're missing _all_ automatic instrumentation since upgrading from .NET Core 2.2 to 3.1 (`mcr.microsoft.com/dotnet/core/aspnet:3.1`). Upgrading APM to 1.14.2 did not help, however all manual instrumentation does come through correctly.

**Comment by ffandreassoroko** [comment]
still the same on 1.16.0

**Comment by ffandreassoroko** [comment]
Hey,
we are now waiting since 4 (or more?) months (since v1.10) that this bug get fixed or getting any kind of help, this is really annoying!
If you need any further information or "tests" to fix that, just ask.

**Comment by lucaspimentel** [comment]
Hey, folks. Thanks for reporting this issue. I sincerely apologize for the delay here. I will try to reproduce this tomorrow. If I can reproduce it quickly (the code samples above should should really help), we can expect a fix soon (possibly in the next release). I'll post my findings here and if needed I'll ask for help reproducing the issue.

**Comment by lucaspimentel** [comment]
Update: I have not been able to reproduce this issue. Here's a [sample app](https://github.com/DataDog/dd-trace-dotnet/tree/lpimentel/missing-aspnetcore-traces/reproductions/AspNetCore31.NoTraces) I wrote to test this and it is generating traces successfully. I tried to keep it as close as possible to the code samples above:
- using Linux containers
- install the tracer using the `deb` package
- also include the `Datadog.Trace` nuget package and replace the default global `Tracer` instance
- minimal app with only custom middleware

Here are some issues to look out for. I hit a few of these myself trying to set this up:
- the default values for Agent settings `DD_APM_ENABLED`, `DD_APM_NON_LOCAL_TRAFFIC`, and `DD_DOGSTATSD_NON_LOCAL_TRAFFIC` have changed over time, or are different in different environments (e.g. Windows vs Linux container), so I like to [set them explicitly](https://github.com/DataDog/dd-trace-dotnet/blob/98d471ffa1e444e83df7ff847be7c7773aabb9bd/reproductions/AspNetCore31.NoTraces/docker-compose.yaml#L8-L10) in the Agent container
- make sure to set the Tracer setting `DD_AGENT_HOST` in the [app container](https://github.com/DataDog/dd-trace-dotnet/blob/98d471ffa1e444e83df7ff847be7c7773aabb9bd/reproductions/AspNetCore31.NoTraces/docker-compose.yaml#L20) so the Tracer knows where to send traces
- if you set the Tracer setting `DD_TRACE_DEBUG` in the [app container](https://github.com/DataDog/dd-trace-dotnet/blob/98d471ffa1e444e83df7ff847be7c7773aabb9bd/reproductions/AspNetCore31.NoTraces/docker-compose.yaml#L21) to write additional logs to `/var/logs/datadog/dotnet` for troubleshooting, you should [create that directory](https://github.com/DataDog/dd-trace-dotnet/blob/98d471ffa1e444e83df7ff847be7c7773aabb9bd/reproductions/AspNetCore31.NoTraces/Dockerfile#L35) before starting the app
- alternatively, [send the logs to stdout](https://github.com/DataDog/dd-trace-dotnet/blob/98d471ffa1e444e83df7ff847be7c7773aabb9bd/reproductions/AspNetCore31.NoTraces/Dockerfile#L36-L37) using symlinks

**Comment by ffandreassoroko** [comment]
Hey, thanks for the answer.

based on your example (and infos how i can enable logs), i was able to get it work. 

its seems that `settings.AgentUri` has no effect
```
var settings = TracerSettings.FromDefaultSources();
settings.AgentUri = new Uri($"http://{AWSEnvironment.DockerHost}:8126");
```
in the logs where the endpoint `http://localhost:8126`, what was wrong, so i added a little bootstrapper script 

```
export DD_AGENT_HOST=$(wget -qO- http://169.254.169.254/latest/meta-data/local-ipv4)

/tmp/dotnet/<backend>
```

tada.... works 👍 

additional, it seems that you createLogPath scripts does not work, after executing nothings happens, folder was not created - had to do it like [you](https://github.com/DataDog/dd-trace-dotnet/blob/98d471ffa1e444e83df7ff847be7c7773aabb9bd/reproductions/AspNetCore31.NoTraces/Dockerfile#L35)

**Comment by lucaspimentel** [comment]
Thanks, @ffandreassoroko. I tried setting the Agent hostname in code the way you did, and had the same result as you. The tracer tries to send traces to `localhost`!

After some debugging, I realized we call `Tracer.Instance.StartDiagnosticObservers()` in our startup hook, before users have a change to replace the default `Tracer.Instance`. The result is that any instrumentation that uses `DiagnosticSource` keeps using that original `Tracer` instance even if the user replaces `Tracer.Instance` with a new instance.

The workaround for now is to not configure settings in code and not replacing the default `Tracer.Instance`. Instead use one of the other configuration method like environment variables.

**Comment by kevin-phillips-ah** [comment]
I've encountered the same issue with tracer versions 1.12.0->1.17.0. I was able to work around the issue using the fix @ffandreassoroko suggested (thanks btw!).

Expected behavior: explicit settings on the `Tracer.Instance` object should apply to all automatic instrumentation.

Observed behavior: ASP.Net integration ignores explicit settings.

We typically rely on the `TracerSettings.AgentUri` property, but observed the following after the application was upgraded from .net core 2.2 to 3.1 (tracer nuget upgrade from 1.9.0 -> 1.12.0 -> 1.17.0). For this test, I stood up an agent locally as 172.17.0.3 to receive traces. Without the environment variable, only HTTP and SQL client automatic instrumentation was sent to correct agent.

```
# without-dd_agent_host-envvar.pcap
root@7638589fd72c:/app# tcpdump -An -vv host 172.17.0.3
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
...
21:37:51.993409 IP (tos 0x0, ttl 64, id 1435, offset 0, flags [DF], proto TCP (6), length 929)
    172.17.0.2.56196 > 172.17.0.3.8126: Flags [P.], cksum 0x5bbb (incorrect -> 0xed5b), seq 1:878, ack 1, win 229, options [nop,nop,TS val 714732323 ecr 1205732096], length 877
E.....@.@...................H.i.....[......
*..#G...POST /v0.4/traces HTTP/1.1
Host: 172.17.0.3:8126
Datadog-Meta-Lang: .NET
Datadog-Meta-Lang-Interpreter: .NET Core
Datadog-Meta-Lang-Version: 3.1.5
Datadog-Meta-Tracer-Version: 1.17.0.0
Datadog-Container-ID: 7638589fd72ca4e7d39c280da676fc15ac17f9d0edc0c71ad6e97b16701ca7a4
x-datadog-tracing-enabled: false
Transfer-Encoding: chunked
Content-Type: application/msgpack
X-Datadog-Trace-Count: 1
1CD
....trace_id.6O^..U.r.span_id.=..&xX.8.name.http.request.resource.,POST <redact>/.service.<redact>-http-client.type.http.start...r...\..duration.'LN..meta..env.dev.s
pan.kind.client.http.method.POST.http.url./https://<redact>/.component.HttpMessageHandler.http-client-handler-type.!System.Net.Http.HttpClientHandler.http.status_code.2
00.metrics.._dd.limit_psr.?........_sampling_priority_v1.?.......
0
21:37:51.993507 IP (tos 0x0, ttl 64, id 46406, offset 0, flags [DF], proto TCP (6), length 52)
    172.17.0.3.8126 > 172.17.0.2.56196: Flags [.], cksum 0x584e (incorrect -> 0x3184), seq 1, ack 878, win 240, options [nop,nop,TS val 1205732155 ecr 714732323], length 0
E..4.F@.@.-V............H.i....B....XN.....
G..;*..#
21:37:51.993888 IP (tos 0x0, ttl 64, id 46407, offset 0, flags [DF], proto TCP (6), length 200)
    172.17.0.3.8126 > 172.17.0.2.56196: Flags [P.], cksum 0x58e2 (incorrect -> 0xea97), seq 1:149, ack 878, win 240, options [nop,nop,TS val 1205732155 ecr 714732323], length 148
E....G@.@.,.............H.i....B....X......
G..;*..#HTTP/1.1 200 OK
Content-Type: application/json
Date: Wed, 17 Jun 2020 21:37:51 GMT
Content-Length: 40
{"rate_by_service":{"service:,env:":1}}
...
21:37:57.036466 IP (tos 0x0, ttl 64, id 1437, offset 0, flags [DF], proto TCP (6), length 52)
    172.17.0.2.56196 > 172.17.0.3.8126: Flags [.], cksum 0x584e (incorrect -> 0x09b6), seq 878, ack 150, win 237, options [nop,nop,TS val 714737367 ecr 1205737155], length 0
E..4..@.@..................BH.i.....XN.....
*...G...
```
and debug logs containing the following
```
# profiler log
2020-06-17 21:37:50.941 +00:00 [DBG] Span closed: [s_id: 4441268478419719224, p_id: null, t_id: 3913450054191795314] for (Service: <redact>-http-client, Resource: POST <redact>/, Operation: http.request, Tags: [env,span.kind,http.method,http.url,component,http-client-handler-type,http.status_code])
{ MachineName: ".", Process: "[1 dotnet]", AppDomain: "[1 <redact>.Web]", TracerVersion: "1.17.0.0" }
2020-06-17 21:37:51.240 +00:00 [DBG] Span closed: [s_id: 4763686930673133264, p_id: null, t_id: 1979386674837199450] for (Service: <redact>.Web, Resource: GET <redact>, Operation: aspnet_core.request, Tags: [component,span.kind,http.method,http.request.headers.host,http.url,language,http.status_code])
{ MachineName: ".", Process: "[1 dotnet]", AppDomain: "[1 <redact_assembly_name_partial>.Web]", TracerVersion: "1.17.0.0" }
2020-06-17 21:37:51.818 +00:00 [ERR] Unable to communicate with the trace agent at "http://localhost:8126/v0.4/traces"
System.Net.Sockets.SocketException (99): Cannot assign requested address
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
```
after restarting the container and specifying the `DD_AGENT_HOST` environment variable, I see the asp.net request in the pcap (snippet for brevity)
```
*.7.G.GvPOST /v0.4/traces HTTP/1.1
Host: 172.17.0.3:8126
Datadog-Meta-Lang: .NET
Datadog-Meta-Lang-Interpreter: .NET Core
Datadog-Meta-Lang-Version: 3.1.5
Datadog-Meta-Tracer-Version: 1.17.0.0
Datadog-Container-ID: c320c3b2a4ae1676971db743bde27c7c57b4c90d02d7e70fd4fca773ad602cef
x-datadog-tracing-enabled: false
Transfer-Encoding: chunked
Content-Type: application/msgpack
X-Datadog-Trace-Count: 1
188
....trace_id.S.F.e.%..span_id.n.n...?..name.aspnet_core.request.resource.GET <redact>.service.<redact>.Web.type.web.start...r..B.L.duration.G.
..meta..component.aspnet_core.span.kind.server.http.method.GET.http.request.headers.host.localhost.http.url.http://<redact>.language.dotnet.http.status_code.200.metrics.._dd.limit_psr
.?........_sampling_priority_v1.?.......
```
Please note that the `TracerSettings.ServiceName` is incorrect and is instead the assembly partial name of the application. I assume by setting the environment variable, this could also be worked around.

**Comment by lucaspimentel** [comment]
This issue is fixed in [release 1.19.5](https://github.com/DataDog/dd-trace-dotnet/releases/tag/v1.19.5). The ASP.NET Core integration will now use settings from `Trace.Instance` even after it is replaced with a new `Tracer` instance. Thanks again for reporting this.
